### ISSUE TITLE
MySQL crashing when faced with rapid directory/notification updates

### ISSUE THREAD

**Issue by kenrestivo** [redmatrix/hubzilla#178]
As part of the bulk importer, I'm doing this:

```php
	proc_run('php','include/notifier.php','location',$channel_id);
	proc_run('php', 'include/directory.php', $channel_id);
```

These lines, when executed in quick succession for a number of channels, appear to cause MySQL to literally crash, segfault and stop running.

I have yet to locate where the dumpfile is and try to pinpoint the error, but I probably will.

This may not be supported use of these calls, and I can possibly schedule the calls with waits, to mitigate this. But I'm noting it here.

**Comment by kenrestivo** [comment]
The query which causes it to fail, according to the mysql log with verbose logging on, is:

```sql
    insert into outq ( outq_hash, outq_account, outq_channel, outq_driver, outq_posturl, outq_async, outq_created, outq_updated, outq_notify, outq_msg ) values (
```

**Comment by royalterra** [comment]
I already see that errors before and now I see some more possible errors can get a crash for many pull request  in the directory link in the login or registration page. Make in Admin the option ON block Public no stop robots, hackers and spammer continues using your web for make pull request using channel suggestions still is open to Public. Anyway my MYSQL problems with Hubzilla Stop when I start using 
Server type: MariaDB
Server version: 10.1.8-MariaDB

**Comment by kenrestivo** [comment]
The query is not specific, it also crashes at other times with:
```sql
	select hubloc_url, hubloc_primary from hubloc where hubloc_addr = 'bamfic@hub.spaz.org' order by hubloc_id desc
```

It appears to be OOM'ing during import (not news to @mike I'm sure). This is a 1GB VM.

**Comment by kenrestivo** [comment]
:(

These fly by at a ridculously fast rate, like microseconds between them:

```apache-log
MySQL server has gone away
2015-11-20 05:53:32:lj0o37foajq788097vsk5tksh0 dba_mysqli.php:35:q: dba_mysqli: ERROR: SELECT id FROM photo WHERE resource_id = '7e5942dfcca72bda5997bc5c932ef295' LIMIT 1
MySQL server has gone away
2015-11-20 05:53:32:lj0o37foajq788097vsk5tksh0 dba_mysqli.php:35:q: dba_mysqli: ERROR: SELECT id FROM photo WHERE resource_id = '851057a65ec5a43452a64941ddb4b6a3' LIMIT 1
MySQL server has gone away
2015-11-20 05:53:32:lj0o37foajq788097vsk5tksh0 dba_mysqli.php:35:q: dba_mysqli: ERROR: SELECT id FROM photo WHERE resource_id = '946b1d25745b2c3bf469080c9d19abfc' LIMIT 1
MySQL server has gone away
2015-11-20 05:53:32:lj0o37foajq788097vsk5tksh0 dba_mysqli.php:35:q: dba_mysqli: ERROR: SELECT id FROM photo WHERE resource_id = '69e2df608e3b7aa8681f6999a4207e15' LIMIT 1
MySQL server has gone away
```

**Comment by kenrestivo** [comment]
@mike OK, I'm lost. What's going on here? 
http://storage.restivo.org/videos/hz-mysql-freakout.ogv

The MySQL hasn't actually gone away in this situation, I can hit it with mysql client, and hubzilla web interface works, but something has gone bananas. Ideas?

**Comment by kenrestivo** [comment]
Backoffs on the importer keep MySQL from crashing. 10 seconds between channel imports, and backoffs exponentially from there, keep things from getting too broken. But it stlll seems very fragile.

**Comment by kenrestivo** [comment]
It appears the cause of the problem is hubzilla's poller cron job. If I turn off the cron job, and stop the polling, I can get through a bulk install without MySQL crashing even once. I don't know what it's doing but it seems to be quite the memory and CPU consumer. It's still a problem though because I'm not sure if users doing a bulk migration will be able to remove the cron job temporarily.

**Comment by kenrestivo** [comment]
Yep, confirmed, it's now a replicable problem. Turning the polling off, no MySQL problems. Manually run the poller, and MySQL explodes into a million flaming shards. Can make it happen at will. This is on a 1GB RAM VM, Debian 8. 

The problem is prompted by importing a large number of channels. That doesnt' crash MySQL in itself, but the next time the poller runs--- BLAMMO.

I don't know what that poller is doing but it seems to be very heavyweight. Ideas, @redmatrix  ?

I'm not sure what the workaround might be for the importer, other than waiting between channel imports some period much larger than the polling period, i.e. 15-20 minutes. Could take a day or two to import a large site, but maybe that's just the way it is.

**Comment by habeascodice** [comment]
I'm not sure we want to run the notifier on imports? People have reported getting notifications about old imported posts as a bug/annoyance. The poller reads outq while the importer is trying to write it, that causes a race condition which causes everything to block on each other, hence the OOM. Similar situation for the directory updates. I'm sure @redmatrix can offer solutions and more thoughts on the subject, but that's why it's happening.

**Comment by kenrestivo** [comment]
This appears to be remedied by #188 . Closing.

**Comment by kenrestivo** [comment]
CPU usage is much improved by #188 but MySQL is stil crashing. It may be possible that further tuning could solve the problem, but that's not been proven by testing yet. Leaving this open until a cause and solution is found.

**Comment by royalterra** [comment]
What I see is today 2 times this process never stop and they start again by cron at the end I have 53 process all duplicates 
www-data	301.85 MB	/usr/bin/php include/queue.php 
At the end  that is eating your CPU I coming tired of holding few hours and I stop mysql and Kill the process after that all look is working perfect again. Why happens that today? That I show there is small but before I have 53 process with 735 megas each

**Comment by royalterra** [comment]
Forget last post make sure you have only in vhost 80 443  -indexes in my server by example Apache 2.4.14 see Apache Vhosts override the option in .htaccess when that is a BUG 
I simple make global -indexes in and CPU everything is now down. Process start and close normally

**Comment by kenrestivo** [comment]
I'd like to know what is the cause of the problem. If that fixes it, great, but until someone can explain why it happens I'm not sure it's a bug in Hubzilla or something that should be documented or what.

**Comment by royalterra** [comment]
By Now is OK will see tomorrow morning if is not over-locked again . I have a web pulling feeds from may urls constantly and making more traffic of Hubzilla to MYSQL and not give this problems . 
For sure is something more because is not normal

**Comment by royalterra** [comment]
Ok I test and restest I not have idea if you have 2 Hubzilla installations in the seam server but after check many logs The problem with php running process start again with more of 1 installation . That is 100% already test 1 domain can be working with many users the other can be a hubzilla installation with not traffic at all. when is activate the cron for both installations The system start holding process open by cron easting you memory and at the end will crash your mysql and you will get MYSQL connection error.
In 6 hours can eat 8 GB RAm and get 85% CPU usage . Still like that I not get MYSQL crash or errors but Apache is Killing my server
